### PR TITLE
Fix #1398 Improve query panel behaviour

### LINKED ISSUE #1398: Improve query panel behaviour
On search collapse the query panel and show only the results.
Change the "Close" button to "Back to search"
Expand the query panel on click to "Back to search"
Limit the minimum height of the results panel to the bottom bar height + header height : the draggable edge should be always visible.

### PR DESCRIPTION
**Pull request by allyoucanmap**
* close drawer menu on search
* close query results on back to search
* close query results on toggle drawer control
* limit the movement of height of query results dock

preview open/close panels:
querypanel

dock minimum and maximum height positions: 
minmaxheight

### CODE REVIEW

File: web/client/components/data/featuregrid/DockedFeatureGrid.jsx [hunk 1]
```diff
@@ -295,6 +297,21 @@ const DockedFeatureGrid = React.createClass({
     selectFeatures(features) {
         this.props.selectAllToggle();
         this.props.selectFeatures(features);
+    },
+    backToSearch() {
+        this.props.onToggleDrawer();
+        this.props.onClose();
+    },
+    limitDockHeight(size) {
+        const minSize = 0.25;
```
Comment (mbarto) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: This should be configurable using a prop

File: web/client/plugins/DrawerMenu.jsx [hunk 2]
```diff
@@ -15,6 +15,8 @@ const {toggleControl, setControlProperty} = require('../actions/controls');
 
 const {changeMapStyle} = require('../actions/map');
 
+const {featureClose} = require('../actions/wfsquery');
```
Comment (mbarto) [inline web/client/plugins/DrawerMenu.jsx]: We are putting querybuilder / featuregrid logic inside a generic container: it's not the right place.
We already have knowledge of querybuilder in the TOC. I don't want to mix stuff too much.

File: web/client/translations/data.fr-FR [hunk 3]
```diff
@@ -693,6 +693,7 @@
             "export": "Exporter",
             "selectall": "Tout sélectionner",
             "deselectall": "Tout désélectionner",
+            "backtosearch": "Back to search",
```
Comment (offtherailz) [inline web/client/translations/data.fr-FR]: Retour à la recherche

File: web/client/translations/data.de-DE [hunk 4]
```diff
@@ -691,6 +691,7 @@
             "export": "Export",
             "selectall": "Alle auswählen",
             "deselectall": "Auswahl aufheben",
+            "backtosearch": "Back to search",
```
Comment (offtherailz) [inline web/client/translations/data.de-DE]: Zurück zur Suche (?)

File: web/client/components/data/featuregrid/DockedFeatureGrid.jsx (hunk 1 above)
Comment (offtherailz) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: should be configurable as a prop

File: web/client/components/data/featuregrid/DockedFeatureGrid.jsx [hunk 5]
```diff
@@ -295,6 +297,21 @@ const DockedFeatureGrid = React.createClass({
...
+        const maxSize = 0.85;
```
Comment (offtherailz) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: Should be configurable, 1 by default

File: web/client/translations/data.de-DE (hunk 4 above)
Comment (offtherailz) [inline web/client/translations/data.de-DE]: Maybe the right place to trigger this action is inside the toggleMenu actionCreator, using thunk to dispatch both actions.

File: web/client/components/data/featuregrid/DockedFeatureGrid.jsx (hunk 1 above)
Comment (allyoucanmap) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: Should they be also action and reducer or only props of the class? (minSize and maxSize)

Comment (offtherailz) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: props of the component and of the plugin (to be customizable in localConfig cfg entry). Pass them from the FeatureGrid Plugin down to this component.

File: web/client/plugins/DrawerMenu.jsx (hunk 2 above)
Comment (offtherailz) [inline web/client/plugins/DrawerMenu.jsx]: Maybe the right place to trigger this action is inside the toggleMenu actionCreator, using thunk to dispatch both actions.

**Comment by allyoucanmap** [comment]
in the previous commit we could back to the TOC menu in two ways from 'back to search' button or
the drawer button but this logic caused problems in DrawerMenu.jsx
before
now we have only one way to back to search menu with the query results 'back to search' button
now

preview new dock limit
newminmaxheight

File: web/client/actions/wfsquery.js [hunk 6]
```diff
@@ -15,6 +15,7 @@ const QUERY_CREATE = 'QUERY_CREATE';
 const QUERY_RESULT = 'QUERY_RESULT';
 const QUERY_ERROR = 'QUERY_ERROR';
 const RESET_QUERY = 'RESET_QUERY';
+const QUERY_DOCK_SIZE = 'QUERY_DOCK_SIZE';
```
Comment (mbarto) [inline web/client/actions/wfsquery.js]: Should we move this to another file? actions/featuregrid?

File: web/client/actions/wfsquery.js [hunk 7]
```diff
@@ -74,6 +75,13 @@ function queryError(error) {
     };
 }
 
+function queryDockSize(dockSize) {
```
Comment (mbarto) [inline web/client/actions/wfsquery.js]: Should we move this to another file? actions/featuregrid?

File: web/client/components/data/featuregrid/DockedFeatureGrid.jsx [hunk 8]
```diff
@@ -54,7 +53,13 @@ const DockedFeatureGrid = React.createClass({
         cleanError: React.PropTypes.func,
         selectAllToggle: React.PropTypes.func,
         zoomToFeatureAction: React.PropTypes.func,
-        onClose: React.PropTypes.func
+        onClose: React.PropTypes.func,
+        openDrawer: React.PropTypes.func,
```
Comment (mbarto) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: onBackToSearch, to replace both openDrawer and enableDrawer: the featuregrid is not aware of drawer existance, and should not be. Each component is independent. We can glue things together in the plugins.

File: web/client/components/data/featuregrid/DockedFeatureGrid.jsx [hunk 9]
```diff
@@ -295,6 +303,20 @@ const DockedFeatureGrid = React.createClass({
     selectFeatures(features) {
         this.props.selectAllToggle();
         this.props.selectFeatures(features);
+    },
+    backToSearch() {
+        this.props.openDrawer();
```
Comment (mbarto) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: this.onBackToSearch()

File: web/client/components/data/query/QueryBuilder.jsx [hunk 10]
```diff
@@ -95,7 +95,8 @@ const QueryBuilder = React.createClass({
                 onQuery: () => {},
                 onReset: () => {},
                 onChangeDrawingStatus: () => {},
-                onToggleDrawer: () => {}
+                closeDrawer: () => {},
+                disableDrawer: () => {}
```
Comment (offtherailz) [inline web/client/components/data/query/QueryBuilder.jsx]: This functions be called into the onQuery action, if needed (using thunk)

File: web/client/components/data/query/QueryToolbar.jsx [hunk 11]
```diff
@@ -103,6 +105,8 @@ const QueryToolbar = React.createClass({
             sortOptions: this.props.sortOptions,
             hits: this.props.hits
         };
+        this.props.actions.closeDrawer();
```
Comment (mbarto) [inline web/client/components/data/query/QueryToolbar.jsx]: closeDrawer and disableDrawer should be handled in onQuery. QueryToolbar does not know it's included in a drawer

File: web/client/components/data/query/QueryToolbar.jsx [hunk 12]
```diff
@@ -104,7 +105,8 @@ const QueryToolbar = React.createClass({
             sortOptions: this.props.sortOptions,
             hits: this.props.hits
         };
-        this.props.actions.onToggleDrawer();
```
Comment (offtherailz) [inline web/client/components/data/query/QueryToolbar.jsx]: same here, there should be an thunk that opens the bottom table and closes the left menu.
Better if it happens when the search result arrives.

File: web/client/reducers/query.js [hunk 13]
```diff
@@ -16,7 +16,8 @@ const {
     QUERY_RESULT,
     QUERY_ERROR,
     RESET_QUERY,
-    FEATURE_CLOSE
+    FEATURE_CLOSE,
+    QUERY_DOCK_SIZE
```
Comment (mbarto) [inline web/client/reducers/query.js]: reducers/featuregrid.js?

File: web/client/reducers/query.js [hunk 14]
```diff
@@ -148,6 +150,11 @@ function query(state = initialState, action) {
                 open: false
             });
         }
+        case QUERY_DOCK_SIZE: {
```
Comment (mbarto) [inline web/client/reducers/query.js]: reducers/featuregrid.js?

File: web/client/plugins/TOC.jsx [hunk 15]
```diff
@@ -117,7 +117,8 @@ const SmartQueryForm = connect((state) => {
             onQuery: createQuery,
             onReset: reset,
             onChangeDrawingStatus: changeDrawingStatus,
-            onToggleDrawer: toggleControl.bind(null, 'drawer', null)
+            closeDrawer: setControlProperty.bind(null, 'drawer', 'enabled', false),
```
Comment (offtherailz) [inline web/client/plugins/TOC.jsx]: for the reasons explained above of course these functions should not be passed here.

File: web/client/components/data/featuregrid/DockedFeatureGrid.jsx (hunk 8 above)
Comment (allyoucanmap) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: Should we include also onClose in onBackToSearch?
```
function closeResponse() {
    return (dispatch, getState) => {
        dispatch(featureClose()); // old onClose
        if (getState().controls && getState().controls.drawer) {
            dispatch(setControlProperty('drawer', 'enabled', true)); // old openDrawer
            dispatch(setControlProperty('drawer', 'disabled', false)); // old enableDrawer
        }
    };
}
```

Comment (mbarto) [inline web/client/components/data/featuregrid/DockedFeatureGrid.jsx]: Yes, I think it's better

**Comment by offtherailz** [comment]
The reason why you disable the drawer menu button is that this way you avoid the user to do things with the layer tree that may make back to search button meanful?

**Comment by allyoucanmap** [comment]
Yes, it's for that reason. I think we should give only one way to the user to come back to the layer tree.
Maybe should we make invisible the drawer button?

**Pull request merged**